src/lib.rs

```diff
@@ -697,76 +633,317 @@ impl<'text> BidiInfo<'text> {
+        return result;
+    }
+
+    // Stop at the lowest *odd* level, since everything below that
+    // is LTR and does not need further reordering
+    min = min.new_lowest_ge_rtl().expect("Level error");
+
+    // For each max level, take all contiguous chunks of
+    // levels ≥ max and reverse them
+    //
+    // We can do this check with the original levels instead of checking reorderings because all
+    // prior reorderings will have been for contiguous chunks of levels >> max, which will
+    // be a subset of these chunks anyway.
+    while min <= max {
+        let mut range = 0..0;
+        loop {
+            range = next_range(levels, range.end, max);
+            result[range.clone()].reverse();
+
+            if range.end >= levels.len() {
+                break;
+            }
+        }
+
+        max.lower(1).expect("Level error");
+    }
+
+    result
+}
+
+/// The core of BidiInfo initialization, factored out into a function that both
+/// the utf-8 and utf-16 versions of BidiInfo can use.
+fn compute_bidi_info_for_para<'a, D: BidiDataSource, T: TextSource<'a> + ?Sized>(
+    data_source: &D,
+    para: &ParagraphInfo,
+    is_pure_ltr: bool,
+    text: &'a T,
+    original_classes: &[BidiClass],
+    processing_classes: &mut [BidiClass],
+    levels: &mut Vec<Level>,
+) where
+    <T as TextSource<'a>>::CharIter: Iterator<Item = char>,
+    <T as TextSource<'a>>::IndexLenIter: Iterator<Item = (usize, usize)>,
+{
+    let new_len = levels.len() + para.range.len();
+    levels.resize(new_len, para.level);
+    if para.level == LTR_LEVEL && is_pure_ltr {
+        return;
+    }
+
+    let processing_classes = &mut processing_classes[para.range.clone()];
+    let levels = &mut levels[para.range.clone()];
+
+    explicit::compute(
+        text,
+        para.level,
+        original_classes,
+        levels,
+        processing_classes,
+    );
+
+    let sequences = prepare::isolating_run_sequences(para.level, original_classes, levels);
+    for sequence in &sequences {
+        implicit::resolve_weak(text, sequence, processing_classes);
+        implicit::resolve_neutral(
+            text,
+            data_source,
+            sequence,
+            levels,
+            original_classes,
+            processing_classes,
+        );
+    }
+    implicit::resolve_levels(processing_classes, levels);
+
+    assign_levels_to_removed_chars(para.level, original_classes, levels);
+}
+
+/// Produce the levels for this paragraph as needed for reordering, one level per *code unit*
+/// in the paragraph. The returned vector includes code units that are not included
+/// in the `line`, but will not adjust them.
+///
+/// This runs [Rule L1], you can run
+/// [Rule L2] by calling [`reorder_visual()`].
+/// If doing so, you may prefer to use [`reordered_levels_per_char()`] instead
+/// to avoid non-character indices.
+///
+/// For an all-in-one reordering solution, consider using [`Self::reorder_visual()`].
```

Comment:
nit: this line and the one above it is not relevant for a private API.

This should just be a low level explanation of the function: what it does (the first line of docs) and what spec stuff it covers (the mention of rule L1)

Reply:
Done in [69aba46](https://github.com/servo/unicode-bidi/pull/109/commits/69aba46d286f4e02a605b1b9612911358a3641c9)